central/reports/scheduler/v2/report_gen.go

```diff
@@ -0,0 +1,326 @@
+package v2
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"sort"
+	"strings"
+	"text/template"
+	"time"
+
+	"github.com/gogo/protobuf/types"
+	"github.com/pkg/errors"
+	"github.com/stackrox/rox/central/reports/common"
+	v1 "github.com/stackrox/rox/generated/api/v1"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/branding"
+	"github.com/stackrox/rox/pkg/errorhelpers"
+	"github.com/stackrox/rox/pkg/mathutil"
+	"github.com/stackrox/rox/pkg/notifiers"
+	"github.com/stackrox/rox/pkg/retry"
+	"github.com/stackrox/rox/pkg/search"
+	"github.com/stackrox/rox/pkg/templates"
+	"github.com/stackrox/rox/pkg/timestamp"
+)
+
+func (s *scheduler) generateAndSendReport(req *ReportRequest, collection *storage.ResourceCollection) error {
+	err := s.upsertReportStatus(req.Ctx, req.ReportMetadata, storage.ReportStatus_PREPARING)
+	if err != nil {
+		return errors.Wrap(err, "Error changing report status to PREPARING")
+	}
+
+	var dataStartTime *types.Timestamp
+	if req.ReportConfig.GetVulnReportFilters().GetSinceLastSentScheduledReport() {
+		dataStartTime, err = s.lastSuccesfulScheduledReportTime(req.Ctx, req.ReportConfig)
+		if err != nil {
+			return errors.Wrap(err, "Error finding last successful scheduled report time")
+		}
+	} else if req.ReportConfig.GetVulnReportFilters().GetSinceStartDate() != nil {
+		dataStartTime = req.ReportConfig.GetVulnReportFilters().GetSinceStartDate()
+	}
```

Comment:
This should be evaluated before report status is changed to `PREPARING`. Ideally, this information should be available in `req` and outside this function.

Reply:
Sounds fair, I moved both dataStartTime and collection to request.